tests/test_onnxml_converter_lightgbm.py

```diff
@@ -0,0 +1,302 @@
+
+if lightgbm_installed():
+    import lightgbm as lgb
+if onnx_installed():
+    import onnxruntime as ort
+if onnx_ml_tools_installed():
+    from onnxmltools.convert import convert_lightgbm
+
+
+class TestONNXConverterLightGBM(unittest.TestCase):
+    def __init__(self, *args, **kwargs):
+        super(TestONNXConverterLightGBM, self).__init__(*args, **kwargs)
+
+    # Base test implementation comparing ONNXML and ONNX models.
+    def _test_lgbm(self, X, model, extra_config={}):
+        # Create ONNX-ML model
+        onnx_ml_model = convert_lightgbm(
+            model, initial_types=[("input", FloatTensorType([X.shape[0], X.shape[1]]))], target_opset=9
+        )
+
+        # Create ONNX model
+        onnx_model = convert(onnx_ml_model, "onnx", X, extra_config)
+
+        # Get the predictions for the ONNX-ML model
+        session = ort.InferenceSession(onnx_ml_model.SerializeToString())
+        output_names = [session.get_outputs()[i].name for i in range(len(session.get_outputs()))]
+        onnx_ml_pred = [[] for i in range(len(output_names))]
+        inputs = {session.get_inputs()[0].name: X}
+        pred = session.run(output_names, inputs)
+        for i in range(len(output_names)):
+            if output_names[i] == "label":
+                onnx_ml_pred[1] = pred[i]
+            else:
+                onnx_ml_pred[0] = pred[i]
+
+        # Get the predictions for the ONNX model
+        session = ort.InferenceSession(onnx_model.SerializeToString())
+        onnx_pred = [[] for i in range(len(output_names))]
+        pred = session.run(output_names, inputs)
+        for i in range(len(output_names)):
+            if output_names[i] == "label":
+                onnx_pred[1] = pred[i]
+            else:
+                onnx_pred[0] = pred[i]
+
+        return onnx_ml_pred, onnx_pred, output_names
+
+    # Utility function for testing regression models.
+    def _test_regressor(self, X, model, rtol=1e-06, atol=1e-06, extra_config={}):
+        onnx_ml_pred, onnx_pred, output_names = self._test_lgbm(X, model, extra_config)
+
+        # Check that predicted values match
+        np.testing.assert_allclose(onnx_ml_pred[0], onnx_pred[0], rtol=rtol, atol=atol)
+
+    # Utility function for testing classification models.
+    def _test_classifier(self, X, model, rtol=1e-06, atol=1e-06, extra_config={}):
+        onnx_ml_pred, onnx_pred, output_names = self._test_lgbm(X, model, extra_config)
+
+        np.testing.assert_allclose(onnx_ml_pred[1], onnx_pred[1], rtol=rtol, atol=atol)  # labels
+        np.testing.assert_allclose(
+            list(map(lambda x: list(x.values()), onnx_ml_pred[0])), onnx_pred[0], rtol=rtol, atol=atol
+        )  # probs
+
+    # Check that ONNXML models can only target the ONNX backend.
+    @unittest.skipIf(
+        not (onnx_ml_tools_installed() and onnx_installed()), reason="ONNXML test require ONNX, ORT and ONNXMLTOOLS"
+    )
+    def test_lightgbm_pytorch(self):
+        X = [[0, 1], [1, 1], [2, 0]]
```

Comment:
Does datasize impact anything we need to test?

Reply:
I think I have a mix of test using random numbers and fixed arrays for that reason. The fixed array tests are taken from the LightGBM convert tests.